### ISSUE TITLE
[3.5.0-dev.20190516] Incorrect type error for mixin

### ISSUE THREAD

**Issue by canonic-epicure** [microsoft/TypeScript#31426]
When using the [mixin pattern](https://www.bryntum.com/blog/the-mixin-pattern-in-typescript-all-you-need-to-know/), there are 2 main notations to define the type of the mixin entity.
The mixin pattern:
```ts
export type AnyFunction<A = any>        = (...input : any[]) => A
export type AnyConstructor<A = object>  = new (...input : any[]) => A
export type Mixin<T extends AnyFunction> = InstanceType<ReturnType<T>>

export const Box = <T extends AnyConstructor<object>>(base : T) =>
class Box extends base {
    value       : any
}
```
1st notation:
```ts
export type Box = Mixin<typeof Box> {}
```
2nd notation:
```ts
export interface Box extends Mixin<typeof Box> {}
```
The 1st notation can not be used for recursive definitions (#29872). Because of that we primarily use 2nd notation. It works fine in most cases, however I found a case, when it produces invalid type error. The full snippet to reproduce the problem below.

Note:
-  The typechecker correctly figures out that there's no `zxc` property on `this`, inside the `observe` method of `Quark` mixin. 
-  In that method, it does not complain about the `this.value` usage
- It does complain, when `value` is used on function argument
- If you'll switch the `Quark` mixin to the 1st notation, the error will disappear

Expected behavior:
- No type errors for the definition of `test` function below

```ts
export type AnyFunction<A = any>        = (...input : any[]) => A
export type AnyConstructor<A = object>  = new (...input : any[]) => A
export type Mixin<T extends AnyFunction> = InstanceType<ReturnType<T>>

export const Box = <T extends AnyConstructor<object>>(base : T) =>
class Box extends base {
    value       : any
}
export interface Box extends Mixin<typeof Box> {}

export const Observable = <T extends AnyConstructor<object>>(base : T) =>
class Observable extends base {
    observe () : Quark {
        return
    }
}
export interface Observable extends Mixin<typeof Observable> {}

export const Quark = <T extends AnyConstructor<Box & Observable>>(base : T) =>
class Quark extends base {

    observe () : Quark {
        // No error here!
        this.value
        
        // error: Error:(28, 14) TS2339: Property 'zxc' does not exist on type 'Quark'.
        this.zxc
        
        return
    }
}
export interface Quark extends Mixin<typeof Quark> {}

const test = (a : Quark) => a.value // <-- Error:(35, 28) TS2339: Property 'value' does not exist on type 'Quark'.
```

**Comment by orta** [comment]
Interesting, so `.value` occurs in the intellisense for a Quark in these examples, perhaps the error is overly greedy here.

**Comment by SamuraiJack** [comment]
Glad to see the progress on this issue!

**Comment by SamuraiJack** [comment]
@orta Perhaps, while you are in the context, you can check why the 2nd notation (#29872) is needed at all? There are some other problems with it, for example, casting a class type to interface loses the static methods information.

**Comment by orta** [comment]
Putting this issue into backlog instead of milestoned, figuring out this issue was first evaluated in https://github.com/microsoft/TypeScript/pull/32770 and then a more permentent  fix was explored in https://github.com/microsoft/TypeScript/pull/33050#issuecomment-525471967 but was considered too expensive on perf for the feature.

**Comment by RyanCavanaugh** [comment]
The `TS2339` error on `a.value` from the original example occurs in TypeScript 3.5.0-dev.20190516, but it no longer occurs in TypeScript 7.1.0-dev.20260902.1. That version instead reports `TS7023` and `TS2310` for the circular `Quark` mixin declaration.

`Quark` extends `Mixin<typeof Quark>`, while the generated class is constrained through `Observable`, whose `observe` method returns `Quark`. Resolving those relationships forms a checking cycle. As described in [Circularity Errors May Occur In The Presence of Circularities](https://github.com/microsoft/TypeScript/wiki/FAQ#circularity-errors-may-occur-in-the-presence-of-circularities), TypeScript may report a circularity error in such cases, so the original expectation of no errors is not guaranteed.
